File: ios/App/Parley/PlaybackController.swift
import AVFoundation
import Foundation
import ParleyKit
import SwiftUI

/// One recording's playback, for one detail screen.
///
/// ## The decode path, and why `AVAudioPlayer` is not it
///
/// Recordings are 16 kHz mono Ogg/Opus — what `OggOpusEncoder` writes and what
/// `PUT /recordings/:id/audio` holds. `AVAudioPlayer` opens that file, reports
/// the right duration and plays it from the top, and for two releases that
/// looked like enough. It is not: **`AVAudioPlayer` cannot seek inside an
/// Ogg/Opus file.** Assigning `currentTime` moves the number it reports back to
/// you and nothing else — decoding carries on from wherever it was — so the
/// scrubber and every "tap a turn to jump there" affordance moved a playhead
/// over audio that never moved with it (#375). Every check written against it
/// passed, because they all asked the player where it was and the answer is
/// exactly the part that lies.
///
/// So the player is gone and `OggPlaybackEngine` (ParleyKit) is here instead:
/// `ExtAudioFile` — the same reader `AudioPeaks` already uses for the waveform,
/// and the one thing that does position correctly in this container — feeding
/// an `AVAudioPlayerNode` → `AVAudioUnitTimePitch` → mixer graph. The time
/// pitch unit is what `enableRate` used to be. `OggPlaybackEngineTests` renders
/// that graph offline and compares the samples against the file, which is the
/// only form of proof this particular bug respects.
///
/// Two Ogg quirks survive the change and still shape the code below:
///
/// - Opening is **not free**. There is no frame count in an Ogg header, so
///   AudioToolbox finds the duration by walking to the last page. For a
///   40-minute meeting that is real work, and it is why the open happens off the
///   main actor behind a `Preparing…` state rather than in a property getter.
/// - `AVAudioFile(forReading:)` opens the same file but reports `length == 0`
///   and then throws on `read`. Do not reach for it here.
///
/// ## What it will not do
///
/// It never starts while the microphone is in use. A live meeting and a playing
/// recording would fight over the audio session, and the loser is always the
/// meeting — the one thing in the app that cannot be repeated. So `play()`
/// checks `MeetingRecorder.holdsTheMicrophone` and the dictation coordinator
/// first and simply does nothing, with nothing shown for it: the play button is
/// a button on a screen you reached *from* the library, and if you are recording
/// you know you are recording.
@MainActor
final class PlaybackController: NSObject, ObservableObject {

    /// Where the audio is, from this screen's point of view. `absent` is the
    /// download model's business, not this object's — the block draws the
    /// download button from `AudioDownloadModel` and only builds a controller
    /// once there is a file.
    enum Phase: Equatable {
        /// No file handed over yet.
        case idle
        /// Opening the file. See the type doc on why this is a state.
        case preparing
        case ready
        case failed(String)
    }

    /// The speed cycle the text button steps through, and the menu's middle.
    static let cycle: [Double] = [1, 1.25, 1.5, 2]
    /// Everything the long-press menu offers.
    static let menu: [Double] = [0.75, 1, 1.25, 1.5, 1.75, 2]
    /// Held by an edge-zone press. Not in the cycle: it is a gesture, not a
    /// setting, and it must not become the persisted rate.
    static let heldRate: Double = 2
    /// The same key `@AppStorage("playbackRate")` would use, so the value is
    /// shared with any view that reads it that way.
    static let rateKey = "playbackRate"

    @Published private(set) var phase: Phase = .idle
    @Published private(set) var isPlaying = false
    @Published private(set) var currentTime: TimeInterval = 0
    @Published private(set) var duration: TimeInterval = 0
    /// The speed the user chose. `2×` while an edge zone is held is *not* this —
    /// see `isHoldingTwoX`.
    @Published private(set) var rate: Double
    @Published private(set) var isHoldingTwoX = false
    /// The overview waveform, or empty while it is still being computed. The
    /// player is usable either way; the waveform draws a flat line until this
    /// arrives.
    @Published private(set) var peaks: [Float] = []

    /// Bumped by every `seek(to:)`.
    ///
    /// A counter rather than a closure or a delegate because the thing that cares
    /// is the transcript's auto-follow, two views away: it resumes on any seek,
    /// and `onChange(of:)` over a counter is how a SwiftUI view observes an
    /// *event* on an object rather than a value. A repeated seek to the same
    /// second still counts as one, which is what a drag needs.
    @Published private(set) var seekGeneration = 0

    private var engine: OggPlaybackEngine?
    /// Which file `engine` holds, so a `.task` that re-runs does not reopen it
    /// and a *different* file does.
    private var loadedURL: URL?
    private var ticker: Task<Void, Never>?
    private var peaksTask: Task<Void, Never>?
    /// Raised from the main actor, polled from the peaks thread between reads —
    /// the computation is synchronous and blocking, so `Task.isCancelled` cannot
    /// reach inside it. Replaced rather than lowered on the next load: a flag
    /// that can go back down is a flag two computations can disagree about.
    private var peaksCancelled = Flag()
    private var sessionActive = false
    private var observers: [NSObjectProtocol] = []
    private let store: LocalAudioStore
    private let recordingId: String

    init(recordingId: String, store: LocalAudioStore = .shared) {
        self.recordingId = recordingId
        self.store = store
        let saved = UserDefaults.standard.double(forKey: Self.rateKey)
        // An unset key reads as 0, which is not a speed.
        self.rate = Self.menu.contains(saved) ? saved : 1
        super.init()
        observeTheSession()
    }

    deinit {
        ticker?.cancel()
        peaksTask?.cancel()
        peaksCancelled.raise()
        for observer in observers { NotificationCenter.default.removeObserver(observer) }
        // The session is handed back here as well as on pause: a screen popped
        // mid-playback must not leave `.playback` active, or the next meeting
        // starts by fighting for a category it should have had for free.
        engine?.close()
        if sessionActive {
            try? AVAudioSession.sharedInstance()
                .setActive(false, options: .notifyOthersOnDeactivation)
        }
    }

    // MARK: opening the file

    /// Open `url` and, once it is playable, start measuring the waveform.
    ///
    /// Idempotent for the file already open — the view calls this from a `.task`
    /// that re-runs on every re-render, and a second call must not restart
    /// playback. A *different* file replaces the current one, which is the
    /// "download removed and fetched again while the screen was open" case.
    func load(url: URL) async {
        guard loadedURL != url else { return }
        unload()
        peaksCancelled = Flag()
        loadedURL = url
        phase = .preparing

        // Off the main actor: see the type doc on what opening an Ogg costs.
        let opened: Result<OggPlaybackEngine, Error> = await Task.detached {
            do {
                return .success(try OggPlaybackEngine(url: url))
            } catch {
                return .failure(error)
            }
        }.value

        // A second `load` for a different file, or an `unload`, can land while
        // the open is in flight. Whatever came back belongs to a screen that has
        // moved on.
        guard loadedURL == url else {
            if case .success(let stale) = opened { stale.close() }
            return
        }

        switch opened {
        case .failure(let error):
            phase = .failed(error.localizedDescription)
        case .success(let engine):
            engine.rate = isHoldingTwoX ? Self.heldRate : rate
            engine.onFinish = { [weak self] in
                Task { @MainActor [weak self] in self?.finish() }
            }
            self.engine = engine
            duration = engine.duration
            phase = .ready
            startPeaks(url: url, seconds: engine.duration)
        }
    }

    /// Back to having no file. Called when the audio is removed from the phone
    /// while the screen is open — the block goes back to offering the download,
    /// and the player must not be left holding a deleted file, an active audio
    /// session, and a duration for something that is gone.
    func unload() {
        engine?.close()
        engine = nil
        loadedURL = nil
        stopTicking()
        releaseSession()
        peaksTask?.cancel()
        peaksCancelled.raise()
        isPlaying = false
        isHoldingTwoX = false
        currentTime = 0
        duration = 0
        peaks = []
        phase = .idle
    }

    /// Cached if it has been computed before, measured off the main actor if not.
    ///
    /// A cache written for a different file — a download removed and replaced
    /// under the same recording id — is spotted by its duration and recomputed
    /// rather than drawn over the wrong timeline.
    private func startPeaks(url: URL, seconds: TimeInterval) {
        if let cached = store.peaks(for: recordingId),
            abs(cached.seconds - seconds) < 0.5
        {
            peaks = cached.peaks
            return
        }
        let id = recordingId
        let store = store
        let cancelled = peaksCancelled
        peaksTask = Task { [weak self] in
            // No `onProgress`. `AudioPeaks.compute` offers it and it is tested,
            // but nothing on this screen would draw it: the waveform's own
            // design is that the player is usable the moment the audio opens and
            // draws a flat line until the peaks land, so a second progress bar
            // would be reporting on something the person is not waiting for.
            let computed = await Task.detached(priority: .utility) {
                try? AudioPeaks.compute(
                    url: url, expectedSeconds: seconds,
                    isCancelled: { cancelled.isRaised })
            }.value
            guard let self, !Task.isCancelled, let computed else { return }
            self.peaks = computed.peaks
            store.putPeaks(computed, for: id)
        }
    }

    // MARK: transport

    /// True when a play tap would do something. False while the microphone is
    /// in use anywhere in the app — see the type doc.
    var canPlay: Bool {
        guard isSeekable else { return false }
        return !MeetingRecorder.holdsTheMicrophone && !DictationCoordinator.shared.active
    }

    /// There is a timeline to move along. True even while the microphone is
    /// busy: a scrub or a timecode tap positions the playhead without asking for
    /// the audio session, so it is allowed where `play()` is not.
    var isSeekable: Bool {
        guard case .ready = phase else { return false }
        return duration > 0
    }

    func play() {
        guard canPlay, let engine else { return }
        activateSession()
        // A file played to the end restarts rather than refusing: the button
        // still says "play", so it has to play.
        if currentTime >= duration - 0.05 { seek(to: 0) }
        engine.rate = isHoldingTwoX ? Self.heldRate : rate
        do {
            try engine.play()
        } catch {
            releaseSession()
            phase = .failed(error.localizedDescription)
            return
        }
        isPlaying = true
        startTicking()
    }

    func pause() {
        engine?.pause()
        isPlaying = false
        stopTicking()
        releaseSession()
    }

    func toggle() {
        isPlaying ? pause() : play()
    }

    /// Move the playhead. Applied live, so a scrub during playback keeps playing
    /// from the new position rather than stopping and resuming.
    ///
    /// Called on every frame of a drag. The engine coalesces — a seek arriving
    /// while the previous one is still re-priming replaces it — so this stays a
    /// cheap call however fast the finger moves.
    func seek(to time: TimeInterval) {
        guard let engine else { return }
        let clamped = min(max(0, time), max(0, duration))
        engine.seek(to: clamped)
        currentTime = clamped
        seekGeneration += 1
    }

    func setRate(_ value: Double) {
        rate = value
        UserDefaults.standard.set(value, forKey: Self.rateKey)
        if !isHoldingTwoX { engine?.rate = value }
    }

    /// The next speed the `1×` button steps to. Off-cycle speeds chosen from the
    /// menu (0.75×, 1.75×) step to the next one *above* them rather than
    /// restarting at 1×, so a tap after a menu choice is still a nudge forwards.
    func cycleRate() {
        let next = Self.cycle.first { $0 > rate } ?? Self.cycle[0]
        setRate(next)
    }

    /// An edge zone is held. Deliberately does not touch `rate`: the chosen
    /// speed has to survive the gesture, because releasing restores it.
    func holdTwoX(_ holding: Bool) {
        guard isHoldingTwoX != holding else { return }
        isHoldingTwoX = holding
        engine?.rate = holding ? Self.heldRate : rate
    }

    /// The end of the file, from the engine's last buffer.
    ///
    /// Parks the playhead at the end rather than snapping it back to zero:
    /// "this is over" is the true state, and the next play tap restarts (see
    /// `play`). The same outcome `audioPlayerDidFinishPlaying` used to produce.
    private func finish() {
        isPlaying = false
        currentTime = duration
        stopTicking()
        releaseSession()
    }

    // MARK: the clock

    /// 10 Hz, and only while playing. The engine's position is a poll — it is
    /// read off the player node's render time — so the choice is a timer or
    /// nothing, and 10 Hz is the slowest rate at which a 1pt playhead crossing a
    /// 350pt waveform still looks continuous.
    private func startTicking() {
        ticker?.cancel()
        ticker = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 100_000_000)
                guard let self, let engine = self.engine, self.isPlaying else { return }
                self.currentTime = engine.currentTime
            }
        }
    }

    private func stopTicking() {
        ticker?.cancel()
        ticker = nil
    }

    // MARK: the audio session

    /// `.playback` for as long as something is playing, and handed straight back
    /// afterwards. Holding it would make this screen the reason the next meeting
    /// cannot open the microphone.
    private func activateSession() {
        guard !sessionActive else { return }
        let session = AVAudioSession.sharedInstance()
        // `.spokenAudio` is the mode for speech the user is listening to: it
        // ducks other audio the way a podcast does rather than mixing.
        try? session.setCategory(.playback, mode: .spokenAudio)
        try? session.setActive(true)
        sessionActive = true
    }

    private func releaseSession() {
        guard sessionActive else { return }
        try? AVAudioSession.sharedInstance()
            .setActive(false, options: .notifyOthersOnDeactivation)
        sessionActive = false
    }

    /// A phone call, Siri, or headphones pulled out.
    ///
    /// Two notifications and no retry ladder. An interruption pauses, and does
    /// not resume itself: the person put the app down to take a call, and a
    /// recording that starts talking again on its own is worse than one that
    /// waits. A configuration change (a route the engine cannot keep its graph
    /// on) is handed to the engine, which rebuilds and picks up where the clock
    /// says it was — `AVAudioPlayer` used to absorb that for us.
    private func observeTheSession() {
        let center = NotificationCenter.default
        observers.append(
            center.addObserver(
                forName: AVAudioSession.interruptionNotification,
                object: AVAudioSession.sharedInstance(), queue: .main
            ) { [weak self] note in
                let raw = note.userInfo?[AVAudioSessionInterruptionTypeKey] as? UInt
                guard let raw, AVAudioSession.InterruptionType(rawValue: raw) == .began else {
                    return
                }
                Task { @MainActor [weak self] in
                    guard let self, self.isPlaying else { return }
                    self.pause()
                }
            })
        observers.append(
            center.addObserver(
                forName: .AVAudioEngineConfigurationChange, object: nil, queue: .main
            ) { [weak self] _ in
                Task { @MainActor [weak self] in
                    guard let self, let engine = self.engine, self.isPlaying else { return }
                    engine.handleConfigurationChange()
                }
            })
    }

    /// A box the blocking peaks computation can poll from another thread.
    private final class Flag: @unchecked Sendable {
        private let lock = NSLock()
        private var raised = false
        var isRaised: Bool {
            lock.lock()
            defer { lock.unlock() }
            return raised
        }
        func raise() {
            lock.lock()
            raised = true
            lock.unlock()
        }
    }
}

/// How a speed is written: `1×`, `1.25×`, `2×`.
///
/// Not a string in the catalog, and deliberately so. It is a number and the
/// multiplication sign, both of which the locale's own number formatting gets
/// right — `0.75×` is `0,75×` in a comma locale for free, and a translator
/// handed `1.25×` to translate would have nothing to do with it.
enum PlaybackRate {
    static func label(_ rate: Double) -> String {
        let number = rate.formatted(
            .number.precision(.fractionLength(0...2)).grouping(.never))
        return "\(number)×"
    }
}

/// `12:34`, or `1:02:03` past the hour. Tabular by convention at every call
/// site — the digits sit under a moving playhead and must not shuffle.
enum PlaybackClock {
    static func string(_ seconds: TimeInterval) -> String {
        let total = Int(seconds.isFinite ? max(0, seconds.rounded(.down)) : 0)
        if total >= 3600 {
            return String(format: "%d:%02d:%02d", total / 3600, (total % 3600) / 60, total % 60)
        }
        return String(format: "%d:%02d", total / 60, total % 60)
    }
}

File: android/app/src/main/kotlin/com/pathors/parley/playback/PlaybackController.kt
package com.pathors.parley.playback

import android.content.Context
import android.util.Log
import androidx.media3.common.util.UnstableApi
import com.pathors.parley.cloud.CloudClient
import com.pathors.parley.cloud.CloudException
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

private const val TAG = "PlaybackController"

/** Where the player is in getting a recording on to the speaker. */
enum class PlaybackPhase {
    /** The audio is not on this phone. The UI offers to fetch it. */
    ABSENT,

    /** `GET /recordings/{id}/audio` is in flight. */
    DOWNLOADING,

    /** The file is here and the engine is opening it. */
    PREPARING,

    /** Playable. This is the only phase that draws a player. */
    READY,

    /** The file is here and will not play, or the download did not finish. */
    FAILED,
}

/** Why playback is not happening. A code — the screen owns the copy for each. */
enum class PlaybackFailure {
    /** The download could not reach the server, or it 5xx'd. */
    DOWNLOAD_NETWORK,

    /** The server has no audio for this recording (404). */
    DOWNLOAD_MISSING,

    /** The file is on the phone and the decoder refused it. */
    UNPLAYABLE,
}

/**
 * Everything the player UI draws, in one immutable snapshot.
 *
 * Position is polled rather than pushed because that is what the engines
 * actually offer, and a 20 Hz poll while playing is cheaper than it looks —
 * nothing recomposes unless a field changed, and the only field that changes
 * between ticks is [positionMs].
 */
data class PlaybackState(
    val phase: PlaybackPhase = PlaybackPhase.ABSENT,
    val failure: PlaybackFailure? = null,
    /** Download progress in 0…1, or -1 when the server declared no length. */
    val downloadFraction: Float = -1f,
    val isPlaying: Boolean = false,
    val positionMs: Long = 0L,
    val durationMs: Long = 0L,
    val rate: Float = 1f,
    /** The overview waveform, once it has been measured. */
    val overview: AudioPeaks.Overview? = null,
    /**
     * Bumped on every [PlaybackController.seekTo]. The transcript watches it to
     * re-arm follow-the-audio: somebody who taps a paragraph to hear it wants
     * the transcript to keep up again, and "the position jumped" is not a thing
     * a position value can say on its own.
     */
    val seekGeneration: Int = 0,
) {
    /** Whether a seek would do anything — i.e. there is audio open. */
    val isSeekable: Boolean get() = phase == PlaybackPhase.READY && durationMs > 0L
}

/**
 * The recording detail screen's player: one recording, from "is the audio even
 * here" to a moving playhead.
 *
 * Owns four things and nothing else: the [PlaybackEngine], the download that
 * feeds it, the position poll, and the overview waveform. It is deliberately
 * not a ViewModel — `RecordingDetailViewModel` holds one and forwards to it, so
 * the playback rules stay readable in one file and testable without Compose.
 *
 * ## Main thread
 *
 * `ExoPlayer` must be built, driven and released on the thread that built it,
 * so every method here hops to [Dispatchers.Main] before touching the engine.
 * The download and the peaks computation are the only work that leaves it.
 *
 * ## Lifetime
 *
 * [release] is mandatory and is called from `onCleared`. A leaked `ExoPlayer`
 * holds an audio focus request and a decoder — on a phone that means a meeting
 * still audibly playing with nothing on screen to stop it.
 */
@androidx.annotation.OptIn(markerClass = [UnstableApi::class])
class PlaybackController(
    context: Context,
    private val cloud: CloudClient,
    private val store: LocalAudioStore,
    private val scope: CoroutineScope,
    /** Swaps the engine for a clock and the audio for a fixture. See [DemoPlaybackEngine]. */
    private val demo: Boolean = false,
) {
    private val context = context.applicationContext

    private val _state = MutableStateFlow(PlaybackState())
    val state: StateFlow<PlaybackState> = _state.asStateFlow()

    private var engine: PlaybackEngine? = null
    private var recordingId: String? = null

    /** The meta's duration, used until the engine reports its own. */
    private var fallbackDurationMs: Long = 0L

    private var ticker: Job? = null
    private var downloadJob: Job? = null
    private var peaksJob: Job? = null
    private var released = false

    /**
     * Point the controller at a recording.
     *
     * Does not download: arriving on a detail screen must not spend somebody's
     * data on a 40 MB file they may only have opened to read. Audio already on
     * the phone opens immediately, which is the common case for a meeting
     * recorded here.
     */
    fun open(recordingId: String, durationMsHint: Long) {
        if (released || this.recordingId == recordingId) return
        this.recordingId = recordingId
        fallbackDurationMs = durationMsHint.coerceAtLeast(0L)

        if (demo) {
            _state.value = PlaybackState(
                phase = PlaybackPhase.READY,
                durationMs = fallbackDurationMs,
                overview = DemoWaveform.overview(recordingId, fallbackDurationMs),
            )
            engine = DemoPlaybackEngine(fallbackDurationMs, engineListener())
            return
        }

        scope.launch {
            val present = withContext(Dispatchers.IO) { store.has(recordingId) }
            if (present) prepare() else _state.value = PlaybackState(phase = PlaybackPhase.ABSENT)
        }
    }

    /**
     * Fetch the audio from the cloud. What the "download to play back" button
     * calls, and a no-op while one is already running.
     */
    fun download() {
        val id = recordingId ?: return
        if (demo || released || downloadJob?.isActive == true) return

        _state.update {
            it.copy(
                phase = PlaybackPhase.DOWNLOADING,
                failure = null,
                downloadFraction = -1f,
            )
        }
        downloadJob = scope.launch { fetchAndPrepare(id) }
    }

    fun togglePlayPause() {
        val engine = engine ?: return
        scope.launch(Dispatchers.Main) {
            if (engine.isPlaying) engine.pause() else engine.play()
            sample()
        }
    }

    /**
     * Move the playhead, and say so.
     *
     * The bump to [PlaybackState.seekGeneration] is the point: it is what
     * re-arms the transcript's follow-the-audio. Clamped here rather than in the
     * engine so a drag past either end of the waveform lands on the end.
     */
    fun seekTo(ms: Long) {
        val engine = engine ?: return
        scope.launch(Dispatchers.Main) {
            val duration = _state.value.durationMs
            val target = if (duration > 0L) ms.coerceIn(0L, duration) else ms.coerceAtLeast(0L)
            engine.seekTo(target)
            _state.update {
                it.copy(positionMs = target, seekGeneration = it.seekGeneration + 1)
            }
        }
    }

    fun setRate(rate: Float) {
        val clamped = rate.coerceIn(RATES.first(), RATES.last())
        val engine = engine
        scope.launch(Dispatchers.Main) {
            engine?.setRate(clamped)
            _state.update { it.copy(rate = clamped) }
        }
    }

    /** Tap-to-cycle through [RATES], wrapping at 2×. */
    fun cycleRate() {
        val current = _state.value.rate
        val index = RATES.indexOfFirst { it > current - RATE_EPSILON && it < current + RATE_EPSILON }
        setRate(RATES[if (index < 0) 1 else (index + 1) % RATES.size])
    }

    /** Mandatory. See the class doc. */
    fun release() {
        released = true
        ticker?.cancel()
        downloadJob?.cancel()
        peaksJob?.cancel()
        val engine = engine ?: return
        this.engine = null
        engine.release()
    }

    // ── internals ────────────────────────────────────────────────────────────

    /**
     * The body of the download job: stream the file down, then open it.
     *
     * Cancellation is rethrown rather than reported, because a cancelled job is
     * the screen going away or a second [download] superseding this one —
     * neither is a failure to show anybody.
     */
    private suspend fun fetchAndPrepare(id: String) {
        try {
            cloud.downloadAudio(id, store.audioFile(id), ::publishDownloadProgress)
            prepare()
        } catch (e: CancellationException) {
            throw e
        } catch (e: Throwable) {
            Log.w(TAG, "audio download failed for $id", e)
            fail(downloadFailure(e))
        }
    }

    /**
     * Report download progress, and only while the download is still the thing
     * happening: the callback runs on the IO thread doing the copy, so a report
     * already in flight when the job is cancelled must not write a progress bar
     * back over whatever replaced it.
     *
     * A total of 0 or less is the server declaring no length; -1 is how
     * [PlaybackState.downloadFraction] says "indeterminate".
     */
    private fun publishDownloadProgress(read: Long, total: Long) {
        val fraction = if (total > 0L) (read.toDouble() / total).toFloat() else -1f
        _state.update { current ->
            if (current.phase == PlaybackPhase.DOWNLOADING) {
                current.copy(downloadFraction = fraction.coerceIn(-1f, 1f))
            } else {
                current
            }
        }
    }

    /**
     * A 404 is the server saying it has no audio for this recording, which is
     * permanent and worth different copy; everything else is treated as a
     * transport problem worth retrying.
     */
    private fun downloadFailure(e: Throwable): PlaybackFailure =
        if ((e as? CloudException)?.isNotFound == true) {
            PlaybackFailure.DOWNLOAD_MISSING
        } else {
            PlaybackFailure.DOWNLOAD_NETWORK
        }

    private suspend fun prepare() {
        val id = recordingId ?: return
        _state.update {
            it.copy(
                phase = PlaybackPhase.PREPARING,
                failure = null,
                durationMs = fallbackDurationMs,
            )
        }
        withContext(Dispatchers.Main) {
            engine?.release()
            engine = try {
                ExoPlaybackEngine(context, store.audioFile(id), engineListener())
            } catch (e: Throwable) {
                Log.w(TAG, "could not open audio for $id", e)
                null
            }
            if (engine == null) fail(PlaybackFailure.UNPLAYABLE)
        }
        loadPeaks(id)
    }

    private fun engineListener() = PlaybackEngineListener(
        onReady = {
            _state.update {
                it.copy(
                    phase = PlaybackPhase.READY,
                    failure = null,
                    durationMs = engine?.durationMs?.takeIf { ms -> ms > 0L }
                        ?: it.durationMs.takeIf { ms -> ms > 0L }
                        ?: fallbackDurationMs,
                )
            }
            // The rate survives a re-open: somebody who set 1.5× and then
            // downloaded a second recording did not ask to go back to 1×.
            engine?.setRate(_state.value.rate)
        },
        onPlayingChanged = { playing ->
            _state.update { it.copy(isPlaying = playing) }
            if (playing) startTicking() else stopTicking()
        },
        onEnded = {
            stopTicking()
            _state.update {
                it.copy(isPlaying = false, positionMs = it.durationMs)
            }
        },
        onFailed = { reason ->
            Log.w(TAG, "player error: $reason")
            fail(PlaybackFailure.UNPLAYABLE)
        },
    )

    private fun fail(failure: PlaybackFailure) {
        stopTicking()
        _state.update {
            it.copy(phase = PlaybackPhase.FAILED, failure = failure, isPlaying = false)
        }
    }

    /**
     * Poll the engine while it is playing.
     *
     * 20 Hz because the waveform's playhead has to move smoothly across it and
     * the transcript has to turn over on the right line; slower reads as a
     * stutter at 2×. It stops the moment playback does, so a paused screen
     * costs nothing.
     */
    private fun startTicking() {
        if (ticker?.isActive == true) return
        ticker = scope.launch(Dispatchers.Main) {
            while (isActive) {
                sample()
                delay(TICK_MS)
            }
        }
    }

    private fun stopTicking() {
        ticker?.cancel()
        ticker = null
    }

    private fun sample() {
        val engine = engine ?: return
        val duration = engine.durationMs.takeIf { it > 0L }
        _state.update {
            it.copy(
                positionMs = engine.positionMs,
                isPlaying = engine.isPlaying,
                durationMs = duration ?: it.durationMs,
            )
        }
    }

    private fun loadPeaks(id: String) {
        peaksJob?.cancel()
        peaksJob = scope.launch {
            val overview = AudioPeaksLoader.load(context, store, id)
            // Guard on the id: a controller that was pointed at a second
            // recording while this was decoding must not be handed the first
            // one's waveform.
            if (recordingId == id) _state.update { it.copy(overview = overview) }
        }
    }

    companion object {
        /** The speed menu, identical to iOS `PlaybackController.menu`. */
        val RATES = listOf(0.75f, 1f, 1.25f, 1.5f, 1.75f, 2f)

        /** Position poll interval while playing. See [startTicking]. */
        private const val TICK_MS = 50L

        private const val RATE_EPSILON = 0.01f
    }
}
